cloudinit/config/cc_wireguard.py

```diff
@@ -0,0 +1,291 @@
+            AllowedIPs = <allowedip1>, <allowedip2>, ...
+        - name: wg1
+          config_path: /etc/wireguard/wg1.conf
+          content: |
+            [Interface]
+            PrivateKey = <private_key>
+            Address = <address>
+            [Peer]
+            PublicKey = <public_key>
+            Endpoint = <endpoint_ip>:<endpoint_ip_port>
+            AllowedIPs = <allowedip1>
+      readinessprobe:
+        - 'systemctl restart service'
+        - 'curl https://webhook.endpoint/example'
+        - 'nc -zv some-service-fqdn 443'
+    """
+        ),
+    ],
+}
+
+__doc__ = get_meta_doc(meta)
+
+LOG = logging.getLogger(__name__)
+
+REQUIRED_WG_INT_KEYS = frozenset(["name", "config_path", "content"])
+WG_CONFIG_FILE_MODE = 0o600
+NL = "\n"
+MIN_KERNEL_VERSION = (5, 6)
+
+
+def supplemental_schema_validation(wg_int: dict):
+    """Validate user-provided wg:interfaces option values.
+
+    This function supplements flexible jsonschema validation with specific
+    value checks to aid in triage of invalid user-provided configuration.
+
+    @param wg_int: Dict of configuration value under 'wg:interfaces'.
+
+    @raises: ValueError describing invalid values provided.
+    """
+    errors = []
+    missing = REQUIRED_WG_INT_KEYS.difference(set(wg_int.keys()))
+    if missing:
+        keys = ", ".join(sorted(missing))
+        errors.append(f"Missing required wg:interfaces keys: {keys}")
+
+    for key, value in sorted(wg_int.items()):
+        if key == "name" or key == "config_path" or key == "content":
+            if not isinstance(value, str):
+                errors.append(
+                    f"Expected a string for wg:interfaces:{key}. Found {value}"
+                )
+
+    if errors:
+        raise ValueError(
+            f"Invalid wireguard interface configuration:{NL}{NL.join(errors)}"
+        )
+
+
+def write_config(wg_int: dict):
+    """Writing user-provided configuration into Wireguard
+    interface configuration file.
+
+    @param wg_int: Dict of configuration value under 'wg:interfaces'.
+
+    @raises: RuntimeError for issues writing of configuration file.
+    """
+    LOG.debug("Configuring Wireguard interface %s", {wg_int["name"]})
+    try:
+        LOG.debug(
+            "Writing wireguard config to file %s", {wg_int["config_path"]}
+        )
+        util.write_file(
+            wg_int["config_path"], wg_int["content"], mode=WG_CONFIG_FILE_MODE
+        )
+    except Exception as e:
+        raise RuntimeError(
+            "Failure writing Wireguard configuration file"
+            f' {wg_int["config_path"]}:{NL}{str(e)}'
+        ) from e
+
+
+def enable_wg(wg_int: dict, cloud: Cloud):
+    """Enable and start Wireguard interface
+
+    @param wg_int: Dict of configuration value under 'wg:interfaces'.
+
+    @raises: RuntimeError for issues enabling WG interface.
+    """
+    try:
+        LOG.debug("Enabling wg-quick@%s at boot", {wg_int["name"]})
```

Comment:
We have some funny-looking formatting throughout, specifically looking at the curly braces.
```
2022-08-03 23:16:42,598 - modules.py[DEBUG]: Running module wireguard (<module 'cloudinit.config.cc_wireguard' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_wireguard.py'>) with frequency once-per-instance
2022-08-03 23:16:42,598 - handlers.py[DEBUG]: start: modules-config/config-wireguard: running config-wireguard with frequency once-per-instance
2022-08-03 23:16:42,598 - util.py[DEBUG]: Writing to /var/lib/cloud/instances/6d9d91a3-a659-41c5-8250-4ced8199065d/sem/config_wireguard - wb: [644] 24 bytes
2022-08-03 23:16:42,598 - helpers.py[DEBUG]: Running config-wireguard using lock (<FileLock using file '/var/lib/cloud/instances/6d9d91a3-a659-41c5-8250-4ced8199065d/sem/config_wireguard'>)
2022-08-03 23:16:42,598 - cc_wireguard.py[DEBUG]: Found Wireguard section in config
2022-08-03 23:16:42,598 - subp.py[DEBUG]: Running command lsmod with allowed return codes [0] (shell=True, capture=True)
2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Configuring Wireguard interface {'wg0'}
2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Writing wireguard config to file {'/etc/wireguard/wg0.conf'}
2022-08-03 23:16:42,601 - util.py[DEBUG]: Writing to /etc/wireguard/wg0.conf - wb: [600] 230 bytes
2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Enabling wg-quick@{'wg0'} at boot
2022-08-03 23:16:42,601 - subp.py[DEBUG]: Running command ['systemctl', 'enable', 'wg-quick@wg0'] with allowed return codes [0] (shell=False, capture=True)
2022-08-03 23:16:42,917 - cc_wireguard.py[DEBUG]: Bringing up interface wg-quick@{'wg0'}
2022-08-03 23:16:42,918 - subp.py[DEBUG]: Running command ['systemctl', 'start', 'wg-quick@wg0'] with allowed return codes [0] (shell=False, capture=True)
2022-08-03 23:16:42,928 - cc_wireguard.py[DEBUG]: Running readinessprobe: '{'curl ifconfig.io > /etc/external.ip'}'
2022-08-03 23:16:42,928 - subp.py[DEBUG]: Running command curl ifconfig.io > /etc/external.ip with allowed return codes [0] (shell=True, capture=True)
2022-08-03 23:16:43,218 - handlers.py[DEBUG]: finish: modules-config/config-wireguard: SUCCESS: config-wireguard ran successfully
```
I'm guessing f-strings got left behind when switching to logging?

Could we fix those throughout the PR please?